### PR TITLE
Signup: DSS: fix invalid markup before storing

### LINKED ISSUE #1890: Signup: DSS: Returning to DSS using Back link causes errors
Steps to reproduce:

1. Visit the DSS flow, eg: http://calypso.localhost:3000/start/dss/
2. Click on a theme; Signup will move to step 2.
3. Click the "Back" link at the bottom of the page to return to step 1.
4. Errors!

screen shot 2015-12-21 at 1 55 12 pm

### PR DESCRIPTION
**Pull request by sirbrillig**
It's possible for invalid markup to be provided as output from the DSS endpoint. This uses the web API to fix markup while it's still a string, before rendering.

Fixes #1890 

## Testing

1. Visit the DSS flow, eg: http://calypso.localhost:3000/start/dss/
2. Click on a theme; Signup will move to step 2.
3. Click the "Back" link at the bottom of the page to return to step 1.
4. Verify the DSS page displays correctly (no sign of weird rendering).
5. Click on another theme.
6. Complete the signup flow.
7. Verify that the site is created and has the theme you chose.

### CODE REVIEW

**Comment by nylen** [comment]
I'm not very familiar with the DSS project.  It looks quite interesting, though.

It looks like the DSS flow currently shows 6 dynamic screenshots, one or more of which have some invalid markup in the underlying screenshot (but the "dynamic" bit is hidden by default).  Which screenshot is it that has the bad markup?  What is the markup that's causing problems, or what is an example of invalid markup that fails here?

I've confirmed that using `cheerio` instead fixes the bug noted at #1890.  However, it also roughly doubles the `signup` bundle size:

```sh
1233451  public-with-cheerio/signup.392113651ead1a988389.js
 683209  public-with-cheerio/signup.392113651ead1a988389.min.js
 540970  public-without-cheerio/signup.521197f0e7a6dd0e6c2e.js
 396352  public-without-cheerio/signup.521197f0e7a6dd0e6c2e.min.js
```

If you need to clean bad markup, it would be better if you can use the DOM parser built in to the browser.  @aduth wrote a method to do this ([`createElementFromString`](https://github.com/Automattic/wp-calypso/blob/INITIAL_COMMIT/client/lib/media-serialization/create-element-from-string.js)).  According to my testing, it works well here:

```js
import createElementFromString from 'lib/media-serialization/create-element-from-string';

function cleanMarkup( markup ) {
    return createElementFromString( markup ).innerHTML;
}
```

The basic approach is to create a `div`, set its HTML, then read it back out.  This was enhanced in 1c215db-gh-calypso-pre-oss to use the [`DOMParser` API](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) if available.

If we can use `createElementFromString` here, we'd probably also want to move it out of `lib/media-serialization` and into somewhere else - `lib/formatting` perhaps?

---

Unrelated to this PR - while testing here, I was observing some bugginess in the code that saves the current step of the flow and restores it on a fresh page load.  Sometimes it worked and sometimes I would be sent backwards 1 step from where I was.

File: client/signup/steps/dss/index.jsx [hunk 1]
```diff
@@ -120,6 +123,22 @@ export default React.createClass( {
...
+				key={ theme.name }
+				themeName={ theme.name }
+				themeSlug={ theme.slug }
+				themeRepoSlug={ 'pub/' + theme.slug }
+				isLoading={ this.state.isLoading }
+				dssImage={ this.state.dssImage }
+				lastSearchTerm={ this.state.lastSearchTerm }
+				markupAndStyles={ this.state.markupAndStyles[ 'pub/' + theme.slug ] }
+				renderComplete={ this.state.renderComplete }
+				{ ...this.props }/>
```
Comment (nylen) [inline client/signup/steps/dss/index.jsx]: I'm in favor of putting the `/>` (and even the `>` for tags with children) on the following line:

- Like `);` and `};` below, it denotes the end of a block
- Easier to rearrange props, like trailing `,` in arrays

That's just my opinion and attempt to change your personal preference :) If you go with the style here, this needs a space in between `}` and `/>`.

Comment (sirbrillig) [inline client/signup/steps/dss/index.jsx]: > I'm in favor of putting the /> (and even the > for tags with children) on the following line,

I agree, but the current .eslintrc disagrees: https://github.com/Automattic/wp-calypso/blob/master/.eslintrc#L61

**Comment by sirbrillig** [comment]
> The basic approach is document.createElement( 'div' ), set its innerHTML, then read it back out. This was enhanced in 1c215db-gh-calypso-pre-oss to use the DOMParser API if available.

That's really cool! OTOH, that's essentially the same approach I tried in #1899 and the argument against it there was that using the DOM to change the string is not compatible with server-side rendering.

**Comment by sirbrillig** [comment]
Swapped out cheerio (that bundle size increase is wayyyy out of scope for this bug fix) for `createElementFromString`. Even if it doesn't work for SSR, it should fix this bug for now and when the next iteration of DSS appears we can figure out how to improve it. (Current DSS test will end tomorrow.)

**Comment by sirbrillig** [comment]
Nevermind. Using `createElementFromString` does not work. It breaks the `Screenshot` component's ability to inject images into the previews:

screen shot 2015-12-22 at 12 56 22 pm

I'm just going to close this for now since it seems like cleaning the markup on the client side is too controversial/complicated and the test ends tomorrow anyway.

**Comment by nylen** [comment]
FWIW, using `innerHTML` in this way **is** compatible with SSR, because you're not setting `innerHTML` for the purpose of rendering.  All you'd need to do is have a library method to clean markup, and have the server-side do the cleaning differently (e.g. using cheerio or jsdom instead of the browser DOM).

But all that is assuming it works :) I guess I didn't unhide the dynamic bits the second time around, just verified that the errors were fixed.

**Pull request closed without merging**